### ISSUE TITLE
attention_mask和padding_mask的问题

### ISSUE THREAD

**Issue by Alwin4Zhang** [wakafengfan/unilm-pytorch#3]
```python
# model_unilm.py

def compute_attention_bias(self, segment_ids): 
        """实现unilm的attention: bert encoder self_attention + look_ahead_mask attention"""
        idxs = torch.cumsum(segment_ids, dim=1)
        mask = idxs[:, None, :] <= idxs[:, :, None]
        mask = mask.to(torch.float).unsqueeze(1)
        mask = -(1.0 - mask) * 1e12
        return mask

    def forward(self, input_ids, token_type_ids):
        extended_attention_mask = self.compute_attention_bias(token_type_ids)
        embedding_output = self.embeddings(input_ids=input_ids, token_type_ids=token_type_ids)
        padding_mask = (input_ids != 0).to(input_ids.device) # padding matrix [1,1,1,1,1,1,0,0,0] # 后面padding补长的直接mask掉
        encoded_layers = self.encoder(embedding_output, attention_mask=extended_attention_mask,
                                      head_mask=padding_mask.unsqueeze(0).expand(self.config.num_attention_heads, token_type_ids.size(0), token_type_ids.size(1)))
        sequence_output = encoded_layers[-1]
        pooled_output = self.pooler(sequence_output)
        return sequence_output, pooled_output
```

```python
...... 
# modeling_bert.py
    attention_scores = attention_scores / math.sqrt(self.attention_head_size)
    if attention_mask is not None:
        # Apply the attention mask is (precomputed for all layers in BertModel forward() function)
        attention_scores = attention_scores + attention_mask   # unilm attention_mask

    # # Normalize the attention scores to probabilities.
    # attention_probs = nn.Softmax(dim=-1)(attention_scores)
    #
    # # This is actually dropping out entire tokens to attend to, which might
    # # seem a bit unusual, but is taken from the original Transformer paper.
    # attention_probs = self.dropout(attention_probs)

    # Mask heads if we want to
    if head_mask is not None:  # padding_mask
        # attention_probs = attention_probs * head_mask
        attention_scores = sequence_masking(attention_scores, head_mask, value='-inf', axis=-1) # padding_mask
......
```
个人理解：作者使用compute_attention_bias实现unilm的attention_mask,head_mask用来实现原生transformers的attention_mask的padding_mask的功能；
有个问题：
输入是[CLS]sentenceA[SEP]sentenceB[SEP]
input_ids在长度不够时会在sentenceB尾部做padding；token_type_ids(也就是segment_ids)会变成`[0,0,0,0,1,1,1,0,0,0]`，就是sentenceA的segment_ids + sentenceB的segment_ids + padding ids；这样做unilm的attention_mask就可以用作者实现方式的先attention_mask + 再利用head_mask实现的padding_mask的方式；
我感觉是不是在前面的unilm attention_mask操作后，sentenceB和padding tokens一起组成的segment_ids`[1,1,1,0,0,0]`在`attention_scores = attention_scores + attention_mask`计算后，这些位置的tokens的attention scores都已经变成-inf(or 负很大的数)，后面直接softmax就变成0了，没必须再进行后面padding mask的操作了？因为padding其实也只有句尾有，句子中也不会有0这个id的token?
